File: M3Tools/Types/Products/Product.cs
﻿using SPPBC.M3Tools.Types.Extensions;

namespace SPPBC.M3Tools.Types
{
    public class Product : DbEntry
    {

        [System.ComponentModel.Browsable(false)]
        [System.Text.Json.Serialization.JsonPropertyName("itemID")]
        public override int Id { get; set; }

        [System.Text.Json.Serialization.JsonPropertyName("itemName")]
        public string Name { get; set; }

        [System.Text.Json.Serialization.JsonPropertyName("stock")]
        public int Stock { get; set; }

        [System.Text.Json.Serialization.JsonPropertyName("price")]
        public decimal Price { get; set; }

        [System.Text.Json.Serialization.JsonPropertyName("available")]
        public bool Available { get; set; }

        public Product() : this(-1)
        {
        }

        public Product(int productID, string name = "New Product", int stock = 0, decimal price = 0m, bool available = false)
        {
            Id = productID;
            Name = name;
            Stock = stock;
            Price = price;
            Available = available;
        }

        public string Display()
        {
            return $"{Id}) {Name} ({Price.FormatPrice()}) {(Available ? "Available" : "Not Available")}";
        }

        public override string ToString()
        {
            return string.Join(My.MySettingsProperty.Settings.ObjectDelimiter, Id, Name, Stock, Price.FormatPrice(), Available ? "Available" : "Not Available");
        }
    }
}

File: M3Tools/Types/Products/Product.vb
﻿Option Strict On
Namespace Types
	Public Class Product
		Inherits DbEntry

		<ComponentModel.Browsable(False)>
		<Text.Json.Serialization.JsonPropertyName("itemID")>
		Public Overrides Property Id As Integer

		<Text.Json.Serialization.JsonPropertyName("itemName")>
		Public Property Name As String

		<Text.Json.Serialization.JsonPropertyName("stock")>
		Public Property Stock As Integer

		<Text.Json.Serialization.JsonPropertyName("price")>
		Public Property Price As Decimal

		<Text.Json.Serialization.JsonPropertyName("available")>
		Public Property Available As Boolean

		Public Sub New()
			Me.New(-1)
		End Sub

		Public Sub New(productID As Integer, Optional name As String = "New Product", Optional stock As Integer = 0, Optional price As Decimal = 0, Optional available As Boolean = False)
			Me.Id = productID
			Me.Name = name
			Me.Stock = stock
			Me.Price = price
			Me.Available = available
		End Sub

		Public Function Display() As String
			Return $"{Id}) {Name} ({Price.FormatPrice}) {If(Available, "Available", "Not Available")}"
		End Function

		Public Overrides Function ToString() As String
			Return String.Join(My.Settings.ObjectDelimiter, Id, Name, Stock, Price.FormatPrice, If(Available, "Available", "Not Available"))
		End Function
	End Class
End Namespace